### PR TITLE
Blessing/Hand of Sacrifice

### PR DESCRIPTION
**Pull request by kadeshar**
- Paladin improvement: Holy paladin now use Blessing/Hand of Sacrifice on tank with low health to mitigate incoming damage

### CODE REVIEW

File: playerbot/strategy/paladin/HolyPaladinStrategy.cpp [hunk 1]
```diff
@@ -46,6 +46,10 @@ void HolyPaladinStrategy::InitCombatTriggers(std::list<TriggerNode*>& triggers)
     triggers.push_back(new TriggerNode(
         "enemy is close",
         NextAction::array(0, new NextAction("judgement", ACTION_NORMAL), NULL)));
+
+    triggers.push_back(new TriggerNode(
+        "blessing of sacrifice",
+        NextAction::array(0, new NextAction("blessing of sacrifice", ACTION_MEDIUM_HEAL + 8), NULL)));
```
Comment (davidonete) [inline playerbot/strategy/paladin/HolyPaladinStrategy.cpp:52]: the triggers are sorted from higher to lower relevance so should be moved up and relevancy should be ACTION_MEDIUM_HEAL + 7

File: playerbot/strategy/paladin/HolyPaladinStrategy.cpp [hunk 2]
```diff
@@ -465,6 +469,10 @@ void HolyPaladinStrategy::InitCombatTriggers(std::list<TriggerNode*>& triggers)
...
```
Comment (davidonete) [inline playerbot/strategy/paladin/HolyPaladinStrategy.cpp:475]: does this make sense to be used when not in combat?

File: playerbot/strategy/paladin/PaladinActions.h [hunk 3]
```diff
@@ -597,6 +597,20 @@ namespace ai
         virtual Value<Unit*>* GetTargetValue() { return context->GetValue<Unit*>("cc target", getName()); }
     };
 
+#ifdef MANGOSBOT_TWO
```
Comment (davidonete) [inline playerbot/strategy/paladin/PaladinActions.h:600]: Macro not needed

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp [hunk 4]
```diff
@@ -367,7 +367,11 @@ namespace ai
                 creators["repentance on snare target"] = &TriggerFactoryInternal::repentance_on_snare_target;
                 creators["repentance interrupt"] = &TriggerFactoryInternal::repentance_interrupt;
                 creators["hammer of justice on enemy"] = &TriggerFactoryInternal::hammer_on_enemy;
-                creators["divine illumination"] = &TriggerFactoryInternal::divine_illumination;
+#ifdef MANGOSBOT_TWO
```
Comment (davidonete) [inline playerbot/strategy/paladin/PaladinAiObjectContext.cpp:370]: macro not needed

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp [hunk 5]
```diff
@@ -434,6 +438,12 @@ namespace ai
             static Trigger* repentance_on_enemy_healer(PlayerbotAI* ai) { return new RepentanceOnHealerTrigger(ai); }
             static Trigger* repentance_on_snare_target(PlayerbotAI* ai) { return new RepentanceSnareTrigger(ai); }
             static Trigger* repentance_interrupt(PlayerbotAI* ai) { return new RepentanceInterruptTrigger(ai); }
+            
+#ifdef MANGOSBOT_TWO
```
Comment (davidonete) [inline playerbot/strategy/paladin/PaladinAiObjectContext.cpp:442]: macro not needed

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp [hunk 6]
```diff
@@ -543,6 +553,11 @@ namespace ai
                 creators["blessing of freedom"] = &AiObjectContextInternal::blessing_of_freedom;
                 creators["avenging wrath"] = &AiObjectContextInternal::avenging_wrath;
                 creators["divine illumination"] = &AiObjectContextInternal::divine_illumination;
+#ifdef MANGOSBOT_TWO
```
Comment (davidonete) [inline playerbot/strategy/paladin/PaladinAiObjectContext.cpp:556]: macro not needed

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp [hunk 7]
```diff
@@ -651,9 +666,6 @@ namespace ai
             static Action* repentance_on_enemy_healer(PlayerbotAI* ai) { return new CastRepentanceOnHealerAction(ai); }
             static Action* holy_shock(PlayerbotAI* ai) { return new CastHolyShockAction(ai); }
             static Action* holy_shock_on_party(PlayerbotAI* ai) { return new CastHolyShockOnPartyAction(ai); }
-            static Action* update_pve_strats(PlayerbotAI* ai) { return new UpdatePaladinPveStrategiesAction(ai); }
```
Comment (davidonete) [inline playerbot/strategy/paladin/PaladinAiObjectContext.cpp:654]: Careful, you have removed this actions, this is important

File: playerbot/strategy/paladin/PaladinTriggers.h [hunk 8]
```diff
@@ -484,6 +484,20 @@ namespace ai
         TurnUndeadTrigger(PlayerbotAI* ai) : HasCcTargetTrigger(ai, "turn undead") {}
     };
 
+#ifdef MANGOSBOT_TWO
```
Comment (davidonete) [inline playerbot/strategy/paladin/PaladinTriggers.h:487]: macro not needed

File: playerbot/strategy/triggers/HealthTriggers.h [hunk 9]
```diff
@@ -29,14 +29,23 @@ namespace ai
...
+		    return ValueInRangeTrigger::IsActive() 
+                && !AI_VALUE2(bool, "dead", GetTargetName())
+                && (!isTankRequired || 
+                    (GetTarget()->IsPlayer() && ai->IsTank((Player*)GetTarget(), false)));
 		}
 
 		virtual float GetValue();
+
+    protected:
+        bool isTankRequired;
```
Comment (davidonete) [inline playerbot/strategy/triggers/HealthTriggers.h:48]: This code is not well tabulated

**Review by davidonete — CHANGES_REQUESTED** [review]

File: playerbot/strategy/triggers/HealthTriggers.h (hunk 9 above)
Comment (kadeshar) [reply playerbot/strategy/triggers/HealthTriggers.h:48]: There is 
obraz

Github messing up.

File: playerbot/strategy/paladin/HolyPaladinStrategy.cpp (hunk 2 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/HolyPaladinStrategy.cpp:475]: Code is in HolyPaladinStrategy::InitCombatTriggers

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp (hunk 7 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/PaladinAiObjectContext.cpp:654]: No i didnt
obraz

File: playerbot/strategy/paladin/HolyPaladinStrategy.cpp (hunk 1 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/HolyPaladinStrategy.cpp]: Sorted

File: playerbot/strategy/paladin/PaladinActions.h (hunk 3 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/PaladinActions.h]: Deleted

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp (hunk 4 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/PaladinAiObjectContext.cpp]: Deleted

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp (hunk 5 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/PaladinAiObjectContext.cpp]: Deleted

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp (hunk 6 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/PaladinAiObjectContext.cpp]: Deleted

File: playerbot/strategy/paladin/PaladinAiObjectContext.cpp (hunk 7 above)
Comment (kadeshar) [reply playerbot/strategy/paladin/PaladinAiObjectContext.cpp:654]: Fixed

**Review by davidonete — APPROVED** [review]

**Pull request merged**